### PR TITLE
Student models & tests for TRILL distillation ICASSP submission

### PR DESCRIPTION
**Pull request by jpeplins**
- Added "MobileNetV3Tiny" Implementation
- Added option to create student model without frontend (In support of distilling on pre-computed spectrograms and targets).
- Added option to average-pool MobileNet output.
- Added option to exclude bottleneck layer.
- Unit tests for above functionality.

### CODE REVIEW

**Comment by jpeplins** [comment]
@googlebot I signed it!

File: non_semantic_speech_benchmark/distillation/models.py [hunk 1]
```diff
@@ -17,42 +17,92 @@
...
       weights=None,
-      pooling=None,
+      pooling='avg' if avg_pool is True else None,
       dropout_rate=0.0)
-  model_out = model(feats)
-  model_out.shape.assert_is_compatible_with([None, 3, 2, 1280])
-  embeddings = tf.keras.backend.batch_flatten(model_out)
-  embeddings.set_shape([None, 3 * 2 * 1280])
+  if avg_pool:
+    model_out = model(feats)
```
Comment (joel-shor) [inline non_semantic_speech_benchmark/distillation/models.py:63]: This line is common, so you can pull this out of the `if` block

File: non_semantic_speech_benchmark/distillation/models.py [hunk 2]
```diff
@@ -17,42 +17,92 @@
...
+  else:
+    embeddings = tf.keras.layers.Flatten(name='distilled_output')(model_out)
   # TODO(joelshor): These final layers can be large. Investigate the compression
   # techniques described in
   # https://blog.tensorflow.org/2020/02/matrix-compression-operator-tensorflow.html?m=1
-  embeddings = tf.keras.layers.Dense(
-      bottleneck_dimension, name='distilled_output')(embeddings)
-  output = tf.keras.layers.Dense(
-      output_dimension, name='embedding_to_target')(embeddings)
+  output = tf.keras.layers.Dense(output_dimension, activation=tf.nn.swish, name='embedding_to_target')(embeddings)
```
Comment (joel-shor) [inline non_semantic_speech_benchmark/distillation/models.py:77]: Why did you add `tf.nn.swish` to the final layer? that seems...wrong, no?

File: non_semantic_speech_benchmark/distillation/models.py [hunk 3]
```diff
@@ -17,42 +17,92 @@
...
+  if avg_pool:
+    model_out = model(feats)
+    model_out.shape.assert_is_compatible_with([None, None])
+  else:
+    model_out = model(feats)
+    model_out.shape.assert_is_compatible_with([None, 3, 2, None])
+  if bottleneck_dimension:
+    embeddings = tf.keras.layers.Flatten()(model_out)
+    embeddings = tf.keras.layers.Dense(
+      bottleneck_dimension, activation=tf.nn.swish, name='distilled_output')(embeddings)
```
Comment (joel-shor) [inline non_semantic_speech_benchmark/distillation/models.py:71]: same here. why?

File: non_semantic_speech_benchmark/distillation/models.py (hunk 1 above)
Comment (jpeplins) [reply non_semantic_speech_benchmark/distillation/models.py:63]: Done.

File: non_semantic_speech_benchmark/distillation/models.py (hunk 2 above)
Comment (jpeplins) [reply non_semantic_speech_benchmark/distillation/models.py:77]: Good catch - see my below comment.

File: non_semantic_speech_benchmark/distillation/models.py (hunk 3 above)
Comment (jpeplins) [reply non_semantic_speech_benchmark/distillation/models.py:71]: This was committed in error. 
Empirically it seems to speed up training (with no observable difference in student quality), but without intuition to justify it.

**Review by joel-shor — APPROVED** [review]
Nice

**Comment by joel-shor** [comment]
Do you know why there's a travis error?

**Comment by jpeplins** [comment]
Travis errors are build issues outside of our project. Seems unrelated to these changes.

**Comment by srjoglekar246** [comment]
Hey Jacob, I see multiple PyLint errors internally, which is blocking submission (for example, 'Access to a protected member _inverted_res_block of a client class' & `lambda x: _parse_examples(x)` -> `_parse_examples`). Can you resolve these (I think yapf might be useful)?

**Comment by jpeplins** [comment]
Hey Sachin, two things:

1. I added a few extra flags to control the learning rate schedule. This is unrelated to the linting issues above.

2. I seemed to have fixed all of the problematic formatting issues. My linter picked up a lot of things that were already committed to the codebase, so I'm assuming those particular issues aren't blocking. To that end, I fixed every _new_ formatting issue my code introduced, with the exception of `Access to a protected member _inverted_res_block`, where I opted to disable the check (using `pylint:disable=protected-access`). This saves me from duplicating a lot of mobilenet code. 

Let me know if you run into any issues.

**Comment by srjoglekar246** [comment]
I submitted Jacob's code with lint changes in [this commit](https://github.com/google-research/google-research/commit/b45380774d8258bceeb5175291eb1e80a3664c3d).

**Comment by joel-shor** [comment]
Excellent! Thanks, Sachin and Jacob.

**Comment by jpeplins** [comment]
Thanks Sachin! I'll close this PR.
